### PR TITLE
Limit number of sessions that are created by same client ip and same user for each graphd

### PR DESCRIPTION
**Pull request by a516072575**
#### What type of PR is this?
- [ ] bug
- [ ] feature
- [x] enhancement

#### What problem(s) does this PR solve?
Issue(s) number: 

Description:
[#3696](https://github.com/vesoft-inc/nebula/pull/3696) provides the ability to delete sessions offline after too sessions are created. It is too late to ensure cluster availability, I think that we should do something to avoid too many session being created. Now there is already max_allowed_connections to limit total number of sessions in cluster to make sure the cluster does not crash because there are too many sessions. But  it is not unfair  to users who correctly use clusters. 
So  I  add max_sessions_per_ip_per_user configuration to limit the number of sessions that are created by same client ip and same user on a graphd service . The default value is 300, You can set it in nebula-graphd.conf file.

#### How do you solve it?

So  I  add max_sessions_per_ip_per_user configuration to limit the number of sessions that are created by same client ip and same user on a graphd service . The default value is 300, You can set it in nebula-graphd.conf file.
  
#### Checklist:
Tests:
- [x] Unit test(positive and negative cases)
- [ ] Function test
- [ ] Performance test
- [ ] N/A

Affects:
- [x] Documentation affected (Please add the label if documentation needs to be modified.)
- [ ] Incompatibility (If it breaks the compatibility, please describe it and add the label.）
- [ ] If it's needed to cherry-pick (If cherry-pick to some branches is required, please label the destination version(s).)
- [ ] Performance impacted: Consumes more CPU/Memory

#### Release notes:

Please confirm whether to be reflected in release notes and how to describe:
> ex. Fixed the bug .....

### CODE REVIEW

**Comment by a516072575** [comment]
Limiting the number of individual user connections is required in a multi tenant cluster. It can prevent one user's mistake from affecting all other users

**Comment by Hera-han** [comment]
@CPWstatic this issue has been open for a long time, please take a look and give feedback. thanks

**Review by CPWstatic — APPROVED** [review]
LGTM.

File: tests/job/test_session.py
Comment (Aiee) [inline tests/job/test_session.py:272]: please add a new line here.

File: src/graph/session/GraphSessionManager.cpp
Comment (Aiee) [inline src/graph/session/GraphSessionManager.cpp:120]: `uiscFindPtr->second.get()->get() + 1 > maxSessions` may cause overflow when the number of session reaches int32_max.
Use `uiscFindPtr->second.get()->get() > maxSessions -1` instead.

Also, please format this code block since it seems to exceed the maximum line width.

File: src/graph/session/GraphSessionManager.cpp
Comment (Aiee) [inline src/graph/session/GraphSessionManager.cpp:324]: `bool GraphSessionManager::addSessionCount(const std::string& key)`

File: src/graph/session/GraphSessionManager.cpp
Comment (Aiee) [inline src/graph/session/GraphSessionManager.cpp:337]: `void GraphSessionManager::subSessionCount(const std::string& key)`

Maybe consider changing the return type to bool to indicate the case that the key is not found?

File: src/graph/session/GraphSessionManager.h
Comment (Aiee) [inline src/graph/session/GraphSessionManager.h:87]: Please use `//` for comments to keep the consistency.

**Review by Aiee — COMMENTED** [review]
Sorry for being back late, and thank you for addressing this problem.
Please take a look at the comments.

**Comment by Aiee** [comment]
LGTM.
Please fix the format reported in https://github.com/vesoft-inc/nebula/runs/5461460167?check_suite_focus=true

**Comment by a516072575** [comment]
> LGTM. Please fix the format reported in https://github.com/vesoft-inc/nebula/runs/5461460167?check_suite_focus=true

I have fixed the format reported, I find that there is not any whitespace at line ending  in code. Why did they report an error?

**Comment by Aiee** [comment]
> > LGTM. Please fix the format reported in https://github.com/vesoft-inc/nebula/runs/5461460167?check_suite_focus=true
> 
> I have fixed the format reported, I find that there is not any whitespace at line ending in code. Why did they report an error?

You could use `clang-format` to format the code. The style config file `.clang-format` is located at the root of the directory.

Ref: https://clang.llvm.org/docs/ClangFormat.html

**Comment by a516072575** [comment]
@Aiee  I have formated code by `clang-format`. Please approve running workflows thanks

**Comment by Aiee** [comment]
> @Aiee I have formated code by `clang-format`. Please approve running workflows thanks

Please check https://github.com/vesoft-inc/nebula/runs/5489976984?check_suite_focus=true
We are using 2 spaces as the indent.

**Comment by a516072575** [comment]
> `[fix the format for all files](/vesoft-inc/nebula/pull/3729/commits/875832d87284c86719bcaa694267088766827368) `

I have formatted all the files involved in this Issue.

**Comment by a516072575** [comment]
> 

@Aiee  I have formatted all the files involved in this Issue.

**Comment by a516072575** [comment]
@Aiee I has fixed the python test case fail, please try to build it again. thx

**Comment by a516072575** [comment]
@Aiee The build failure looks not related to my code

File: tests/job/test_session.py
Comment (Aiee) [inline tests/job/test_session.py:272]: Please fix the new like here

**Review by Aiee — COMMENTED** [review]
Well done!

File: tests/job/test_session.py
Comment (a516072575) [reply tests/job/test_session.py:272]: what is meaning？delete last line ?

Comment (Aiee) [reply tests/job/test_session.py:272]: You may use  https://stackoverflow.com/questions/44704968/visual-studio-code-insert-new-line-at-the-end-of-files as a reference if you are using vs code. And don't forget to sign the CLA so we can merge this PR.

Comment (a516072575) [reply tests/job/test_session.py:272]: So I don't need to do anything anymore before this PR is merged

Comment (Aiee) [reply tests/job/test_session.py:272]: You may check line 272 to see if there is any space. The file should end with a new line without spaces.
You also need to sign the CLA https://cla-assistant.io/vesoft-inc/nebula?pullRequest=3729 for this PR can be merged.

**Review by Aiee — APPROVED** [review]
Good job! Thank you for your contribution.

**Comment by a516072575** [comment]
> Good job! Thank you for your contribution.

Thank you for your patience and help

**Comment by Aiee** [comment]
> > Good job! Thank you for your contribution.
> 
> Thank you for your patience and help

You are welcome! :D We look forward to seeing you in the community again!
BTW, you need to add  冯亮(26011345) to your GitHub account so that all committers could sign the CLA
Screen Shot 2022-03-10 at 18 43 05

**Comment by a516072575** [comment]
> BTW, you need to add 冯亮(26011345) to your GitHub account so that all committers could sign the CLA
`冯亮(26011345) ` is my company's system account, I forget to change the Git configuration when I sometimes submitted.


> > > Good job! Thank you for your contribution.
> > 
> > 
> > Thank you for your patience and help
> 
> You are welcome! :D We look forward to seeing you in the community again! BTW, you need to add 冯亮(26011345) to your GitHub account so that all committers could sign the CLA Screen Shot 2022-03-10 at 18 43 05

`冯亮(26011345)` is my system account in my company  , I forget to change the Git configuration when I  submitted. It is not a github account, can we ignore it?

**Comment by Aiee** [comment]
> 

Unfortunatlly, it is 

> > BTW, you need to add 冯亮(26011345) to your GitHub account so that all committers could sign the CLA
> > `冯亮(26011345) ` is my company's system account, I forget to change the Git configuration when I sometimes submitted.
> 
> > > > Good job! Thank you for your contribution.
> > > 
> > > 
> > > Thank you for your patience and help
> > 
> > 
> > You are welcome! :D We look forward to seeing you in the community again! BTW, you need to add 冯亮(26011345) to your GitHub account so that all committers could sign the CLA Screen Shot 2022-03-10 at 18 43 05
> 
> `冯亮(26011345)` is my system account in my company , I forget to change the Git configuration when I submitted. It is not a github account, can we ignore it?

Unfortunately, it is mandatory to sign the CLA. You could [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user) and may remove your working email later.

**Comment by a516072575** [comment]
@Aiee 
This problem has been resolved
image

**Review by CPWstatic — APPROVED** [review]

**Comment by Aiee** [comment]
@a516072575  Please take a look at the failed pytest case

**Comment by Sophie-Xie** [comment]
> @a516072575 Please take a look at the failed pytest case. You could run the test locally before pushing the commit.

https://github.com/vesoft-inc/nebula/runs/5533203149?check_suite_focus=true

E   nebula3.Exception.AuthFailedException: b"Create Session failed: Too many sessions created from 127.0.0.1 by user root. the threshold is 1. You can change it by modifying 'max_sessions_per_ip_per_user' in nebula-graphd.conf"

**Comment by CPWstatic** [comment]
Please add `max_sessions_per_ip_per_user` to `nebula-graphd.conf.production` file.

**Comment by a516072575** [comment]
> E nebula3.Exception.AuthFailedException: b"Create Session failed: Too many sessions created from 127.0.0.1 by user root. the threshold is 1. You can change it by modifying 'max_sessions_per_ip_per_user' in nebula-graphd.conf"

I will complete it as soon as possible

**Comment by a516072575** [comment]
> > @a516072575 Please take a look at the failed pytest case. You could run the test locally before pushing the commit.
> 
> https://github.com/vesoft-inc/nebula/runs/5533203149?check_suite_focus=true
> 
> E nebula3.Exception.AuthFailedException: b"Create Session failed: Too many sessions created from 127.0.0.1 by user root. the threshold is 1. You can change it by modifying 'max_sessions_per_ip_per_user' in nebula-graphd.conf"

I  will fix it today

**Review by CPWstatic — APPROVED** [review]

**Review by Aiee — APPROVED** [review]
We made it!

**Pull request merged**
